### ISSUE TITLE
Event based callback signatures

### ISSUE THREAD

**Issue by mnesarco** [mnesarco/fcapi#15]
@galou 
I will introduce a different approach here to make it more flexible so i don't have to change the signatures again in the future, I want to do this change once and only once. So I will introduce Event classes everywhere and all signatures will receive a specific Event and no more arguments. this way Event classes can be enhanced over time without breaking existing code.

Example (pseudocode):

```python

class PropertyChangeEvent[T]:
  old_value: T
  new_value: T
  target: DocumentObject


@my_prop.observer
def listener3(self, event: PropertyChangeEvent):
  print(f"my_property3 has changed to {event.new_value}")


```

I will introduce the same approach for everything `on_execute`, `on_init`, ....

This also makes signatures easier to remember.

_Originally posted by @mnesarco in https://github.com/mnesarco/fcapi/issues/2#issuecomment-2524018855_

**Comment by galou** [comment]
I cannot assess whether this will render the API more developer-friendly. It's probably a good idea but it appears to me this may not be necessary. The signature for `on_execute` is `(self)`, this cannot be easier to remember. Anyway, you're on the best position to assess and I like the fact that there's no redundancy anymore.

**Comment by mnesarco** [comment]
@galou The main goal is to make the API stable, so data passed to the api methods can change over time without breaking existing code. The evolution of FreeCAD API is unpredictable that is why I don't use inheritance (that would be easier) but metaprogramming to adapt to the changes internally while maintains a stable and documented api exposed to the extension developers. I think this change is worth to do now that not too many people uses fcapi, then when more and more extensions depend on it, it will be harder and harder to introduce changes.

With this change, all methods can be implemented with one of two signatures, for example

```python
def on_execute(self):
  # Valid
  
def on_execute(self, event):
  # Valid
  
def on_execute(self, event: ExecuteEvent):
  # Valid
  
```

**Comment by mnesarco** [comment]
CC @luzpaz I don't know how familiar are you with current fcapi but you have been following my previous scripting related work. Do you want to comment something about this planned change?

**Comment by tomate44** [comment]
Hi,
I have only started using fcapi a few days ago, and I am far less skilled than you, so I fully trust you and will adapt to your decision.

**Comment by galou** [comment]
With the flexible signature it makes a lot of sense.

**Comment by luzpaz** [comment]
Thanks @mnesarco but unfortunately this is out of my purview. Maybe worth running by wwmayer ?

**Comment by mnesarco** [comment]
I believe the change is worth to do. I will push it to the devel branch first.

**Comment by galou** [comment]
Nice, let met know when I can try out.

**Comment by mnesarco** [comment]
@galou you can download the file from the **devel** branch to try it out:
https://github.com/mnesarco/fcapi/blob/devel/fpo.py

but documentation is still outdated.

**Comment by mnesarco** [comment]
Hello Friends, Any comments about the proposed changes in the devel branch? I plan to go with it this week and update docs...

**Comment by galou** [comment]
Sorry, I didn't have the time to test it out.

**Comment by mnesarco** [comment]
As @galou has been reportedly using the `devel` branch and I am also using it in my projects, sounds like the Event based signatures is good enough to merge in main. The blocking issue is that I need to update the documentation.

**Comment by mnesarco** [comment]
merged into main
